Prusa-CW-Firmware/main.cpp

```diff
@@ -2090,55 +2236,61 @@ SIGNAL(TIMER0_COMPA_vect) //1ms timer
     read_encoder();
   }
 
-  if (heater_running) {
-    fan_heater_rpm();
-  }
-  else
-  {
-    ams_counter = 0;
-  }
-
   if (pid_mode) {
-    if (curing_machine_mode == 0 || curing_machine_mode == 2) {
-      if (chamber_temp_celsius >= target_temp_celsius) {
-        actualTemp = chamber_temp_celsius;
-        targetTemp = target_temp_celsius;
-        PID();
-        fan1_duty = newSpeed;
-        fan2_duty = newSpeed;
-      }
-      else {
-        fan1_duty = FAN1_MENU_SPEED;
-        fan2_duty = FAN2_MENU_SPEED;
-      }
-    }
-    if (curing_machine_mode == 3) {
-      if (chamber_temp_celsius >= resin_target_temp_celsius) {
-        actualTemp = chamber_temp_celsius;
-        targetTemp = resin_target_temp_celsius;
-        PID();
-        fan1_duty = newSpeed;
-        fan2_duty = newSpeed;
-      }
-      else {
-        fan1_duty = FAN1_MENU_SPEED;
-        fan2_duty = FAN2_MENU_SPEED;
-      }
+    byte tmpTargetTemp;
+    if (config.curing_machine_mode == 0 || config.curing_machine_mode == 2 || config.curing_machine_mode == 3 || (selftest.phase == 5 && state == SELFTEST)) {
+    	if(config.curing_machine_mode != 3)
+    		tmpTargetTemp = config.target_temp_celsius;
+    	else
+    		tmpTargetTemp = config.resin_target_temp_celsius;
+
+    	if (chamber_temp_celsius >= tmpTargetTemp) {
+    		fan_duty[0] = PID(chamber_temp_celsius, tmpTargetTemp);
+    		fan_duty[1] = fan_duty[0];
+    	} else {
+    		fan_duty[0] = FAN1_MENU_SPEED;
+    		fan_duty[1] = FAN2_MENU_SPEED;
+    	}
     }
   }
 
   fan_pwm_control();
 
   fan_rpm();
 
+  if(state == SELFTEST && ms_counter % 1000 == 0){
+	  if(selftest.phase == 3 || selftest.phase == 4 || selftest.phase == 5){
+		  if(selftest.phase == 3 && selftest.vent_test != true){
+			  lcd.setCursor(8, 1);
+			  lcd.print(selftest.fan_tacho[0]);
+			  lcd.setCursor(8, 2);
+			  lcd.print(selftest.fan_tacho[1]);
+	      }
+		  if(selftest.phase == 5 && selftest.heater_test != true){
+			  lcd.setCursor(8, 1);
+			  lcd.print(chamber_temp_celsius, 1);
+		  }
+	      byte lcd_min = selftest.tCountDown.getCurrentMinutes();
+	      byte lcd_sec = selftest.tCountDown.getCurrentSeconds();
+	      if(lcd_min > 9)
+	    	  lcd.setCursor(6, 3);
+	      else
+	    	  lcd.setCursor(7, 3);
+	      lcd.print(lcd_min);
+	      lcd.print(":");
+	      if(lcd_sec < 10)
+	    	  lcd.print("0");
+	      lcd.print(lcd_sec);
```

Comment:
It is generally dangerous to use some periphery (lcd) from main loop and from interrupt routine.

Reply:
Also are you sure that such complex code is executed in less than 1ms? Otherwise you will miss some 1ms ticks.